Add liquidation actor tests

luckysori opened the pull request:
After the introduction of the `itchysats/order` protocol we can also prove that the CFD includes liquidation events from the moment it is open (not the case when we first merged https://github.com/itchysats/itchysats/pull/2378).

When adding these tests we discovered and fixed a bug which was causing us to only generate _short_ liquidation events.

---

I'm gonna create an issue to add more tests based on the `maker_position`, because that determines the shape of the payout curve, which affects the liquidation intervals.

Interestingly, here (maker going short) I only check the _long_ liquidation events, because the short liquidation events are pretty useless and hard to test: we have CETs for one specific price (as opposed to an interval for long liquidation CETs). This is because of the way our payout curve is implemented.

luckysori committed:
Refactor start_from_open_cfd_state test helper

luckysori committed:
Use From impl instead of Decimal::from_u64

`Decimal::from_u64` is a trait implementation that has to conform to
the type signature of the trait, but it actually always returns
`Some(decimal)`.

luckysori committed:
Add liquidation actor tests

After the introduction of the `itchysats/order` protocol we can also
prove that the CFD includes liquidation events from the moment it is
open (not the case when we first merged
https://github.com/itchysats/itchysats/pull/2378).

When adding these tests we discovered and fixed a bug which was
causing us to only generate _short_ liquidation events.

luckysori committed:
Simplify open_cfd test helper

The `FeeCalculator` is not used in `open_cfd` and can be built before
calling it.

We also verify that when opening a CFD we choose a `taker_leverage`
which is actually supported by the `leverage_choices` configured in
the `OfferParams`. And that we use the expected `taker_leverage` when
calling `place_order`!

da-kami commented on daemon-tests/tests/liquidation.rs:
```diff
@@ -0,0 +1,161 @@
+        )
+    );
+
+    taker.system.commit(order_id).await.unwrap();
```
🙃 Given that from here both tests do the same we could also extract that into a `force_close` function. This might help us when adapting the code so we don't miss a spot. Not super important given only one duplication so far :)

luckysori replied:
Thanks for suggesting this, I hadn't considered it!

da-kami commented on daemon-tests/tests/liquidation.rs:
```diff
@@ -0,0 +1,161 @@
+
+    let oracle_data = OliviaData::example_0();
+
+    // We set the initial price to a value much higher than that of the price that the mock oracle
```
Nifty! Took me a bit to understand the comment but makes total sense :)

luckysori replied:
It's quite hard to explain :( Maybe there's a different way to write this code so that it doesn't require such a complex comment.

da-kami replied:
I thought about it when I had a look, but did not have a better idea so far 😅
I think it's fine :)

da-kami commented on daemon-tests/tests/liquidation.rs:
```diff
@@ -0,0 +1,161 @@
+    let random_liquidation_event = *taker
+        .latest_dlc()
+        .liquidation_event_ids()
+        .choose(&mut thread_rng())
```
Could we potentially check all or at least some deterministic values that represent sane boundaries? 
I would like to avoid randomness when possible in the tests to avoid flaky tests.

luckysori replied:
Fair point. I think checking all of them is a no-go because it either would take too long (24 times longer and even more once we have more frequent liquidation events) or require hacky tricks like erasing/overwriting events from the test DB.

Perhaps the lesson here is that we should not try to capture everything with this actor test. I will default back to just checking the first liquidation event. And I will consider adding a unit test somewhere else that verifies that all liquidation events are correct.

da-kami reviewed (approved):
Nice! Great the test helped to detect a problem :)

luckysori committed:
Take &Attestation when simulating in mocked oracle actor

luckysori committed:
Extract settle_non_collaboratively test helper

luckysori committed:
Create payout_curve::prop_compose test submodule

So that the `proptest::Strategy`s can be reused in other modules.

luckysori committed:
Prove that the payout curve intervals have no gaps

By adding a proptest.

luckysori committed:
Document payout curve bug

Payout curve price intervals are expected to be monotonically
increasing. That means that they should never go backwards. If they
did, they would overlap with previous intervals. That would cause more
than one CET to be activated when the oracle attests to a price in the
overlapping intervals.

In fact, if this happens we would fail when calling
`maia_core::generate_payouts` during contract setup or rollover.

We add a proptest that fails, but mark it with the `ignore` attribute.
The bug seems to occur depending on the `initial_price`,
`long_leverage`, `short_leverage` and `n_payouts`. The other arguments
appear to have no effect on whether we run into this bug or not.

luckysori committed:
Prove that OraclePayouts can be built from Payouts

The scope of the proptest added is limited by the bug documented in
the parent commit.

Nonetheless, if we choose a set of values that don't run into the
aforementioned bug (unrelated to the code we're testing here), we
demonstrate that the property holds: that we can build `OraclePayouts`
from `Payouts` with the expected number of events and the expected
number of liquidation payouts (2 per event).

da-kami reviewed (approved):
Great! Love the proptests!

Would it be worth adding a ticket to investigate the documented bug further?
